gapis/api/vulkan/transform_capture_log.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (C) 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vulkan
+
+import (
+	"context"
+	"os"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/core/memory/arena"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/capture"
+)
+
+type captureLog struct {
+	file   *os.File
+	header *capture.Header
+	cmds   []api.Cmd
+	ids    map[api.CmdID]api.CmdID
+}
+
+func newCaptureLog(ctx context.Context, sourceCapture *capture.GraphicsCapture, path string) *captureLog {
+	f, err := os.Create(path)
+	if err != nil {
+		log.E(ctx, "Failed to create replay capture file %v: %v", path, err)
+		return nil
+	}
+	return &captureLog{
+		file:   f,
+		header: sourceCapture.Header,
+		cmds:   []api.Cmd{},
+		ids:    map[api.CmdID]api.CmdID{},
+	}
+}
+
+func (logTransform *captureLog) RequiresAccurateState() bool {
+	return false
+}
+
+func (logTransform *captureLog) BeginTransform(ctx context.Context, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	return inputCommands, nil
+}
+
+func (logTransform *captureLog) ClearTransformResources(ctx context.Context) {
+	// Do nothing
+}
+
+func (logTransform *captureLog) TransformCommand(ctx context.Context, id api.CmdID, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	// Melih TODO: Find a way to filter the custom commands
+	// Don't write out the custom commands (e.g. replay.Custom)
+	// if cmd.API() != nil {
+	// 	if id.IsReal() {
+	// 		t.ids[id] = api.CmdID(len(t.cmds))
+	// 	}
+	// 	t.cmds = append(t.cmds, cmd)
+	// }
+
+	if len(inputCommands) > 0 {
+		logTransform.ids[id] = api.CmdID(len(logTransform.cmds))
+		logTransform.cmds = append(logTransform.cmds, inputCommands...)
+	}
+
+	return inputCommands, nil
+}
+
+func (logTransform *captureLog) EndTransform(ctx context.Context, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	a := arena.New()
+	defer a.Dispose()
+
+	for idx := range logTransform.cmds {
+		// Melih TODO: Is this clonning still required?
+		// DCE clones it as well before creating a new capture.
```

Comment:
I'm not sure. Can anyone else answer this?

Reply:
Maybe andrew but not necessary. This only runs with a special flag. No need for any optimisation.